File: pwa/src/util/distance-utils.ts
/**
 * Haversine distance and initial bearing calculations.
 * Ported from DistanceUtils.kt in the Android app.
 */

const EARTH_RADIUS_METERS = 6371000;

/** Degrees to radians. */
function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

/** Radians to degrees. */
function toDeg(rad: number): number {
  return (rad * 180) / Math.PI;
}

/**
 * Calculate distance in meters between two WGS84 points
 * using the Haversine formula.
 */
export function distanceMeters(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;
  return EARTH_RADIUS_METERS * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/**
 * Calculate initial bearing (0=north, clockwise) from point 1 to point 2.
 * Returns degrees in range [0, 360).
 */
export function bearingDegrees(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const phi1 = toRad(lat1);
  const phi2 = toRad(lat2);
  const dLambda = toRad(lon2 - lon1);
  const y = Math.sin(dLambda) * Math.cos(phi2);
  const x =
    Math.cos(phi1) * Math.sin(phi2) -
    Math.sin(phi1) * Math.cos(phi2) * Math.cos(dLambda);
  return (toDeg(Math.atan2(y, x)) + 360) % 360;
}

/**
 * Format distance for display: meters if <1km, km with one decimal otherwise.
 */
export function formatDistance(meters: number): string {
  if (meters < 1000) {
    return `${Math.round(meters)} m`;
  }
  return `${(meters / 1000).toFixed(1)} km`;
}

File: app/src/main/java/no/naiv/tilfluktsrom/util/DistanceUtils.kt
package no.naiv.tilfluktsrom.util

import kotlin.math.*

/**
 * Distance and bearing calculations using the Haversine formula.
 */
object DistanceUtils {

    private const val EARTH_RADIUS_METERS = 6371000.0

    /**
     * Calculate the distance in meters between two WGS84 points.
     */
    fun distanceMeters(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val dLat = Math.toRadians(lat2 - lat1)
        val dLon = Math.toRadians(lon2 - lon1)
        val a = sin(dLat / 2).pow(2) +
            cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) *
            sin(dLon / 2).pow(2)
        return EARTH_RADIUS_METERS * 2 * atan2(sqrt(a), sqrt(1 - a))
    }

    /**
     * Calculate the initial bearing (in degrees, 0=north, clockwise)
     * from point 1 to point 2.
     */
    fun bearingDegrees(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val phi1 = Math.toRadians(lat1)
        val phi2 = Math.toRadians(lat2)
        val dLambda = Math.toRadians(lon2 - lon1)
        val y = sin(dLambda) * cos(phi2)
        val x = cos(phi1) * sin(phi2) - sin(phi1) * cos(phi2) * cos(dLambda)
        return (Math.toDegrees(atan2(y, x)) + 360) % 360
    }

    /**
     * Format distance for display: meters if <1km, otherwise km with one decimal.
     */
    fun formatDistance(meters: Double): String {
        if (meters.isNaN()) return "—"
        return if (meters < 1000) {
            "${meters.toInt()} m"
        } else {
            "${"%.1f".format(meters / 1000)} km"
        }
    }
}
